genson/src/main/java/com/owlike/genson/reflect/EvolvableHandler.java

```diff
@@ -0,0 +1,63 @@
+package com.owlike.genson.reflect;
+
+import com.owlike.genson.Context;
+import com.owlike.genson.Converter;
+import com.owlike.genson.ext.jsr353.JSR353Bundle;
+import com.owlike.genson.stream.ObjectReader;
+import com.owlike.genson.stream.ObjectWriter;
+
+import javax.json.JsonValue;
+
+import java.util.Map;
+import java.util.function.Consumer;
+
+/**
+ * An implementation of an {@link UnknownPropertyHandler} that supports
+ * evolution of data classes via {@link Evolvable} interface.
+ * <p>
+ * If the target object we are deserializing into is {@link Evolvable},
+ * this handler will add any unknown properties encountered during
+ * deserialization into {@link Evolvable#unknownProperties()} map,
+ * and will write them out along with all known properties during
+ * subsequent serialization.
+ * <p>
+ * This prevents data loss when serializing and deserializing the same
+ * JSON payload using different versions of Java data classes.
+ *
+ * @author Aleksandar Seovic  2018.05.20
+ */
+public class EvolvableHandler implements UnknownPropertyHandler {
+    private static final Converter<JsonValue> CONVERTER = new JSR353Bundle.JsonValueConverter();
```

Comment:
The only problem is that this assumes we are in Java EE world and I'd like Genson to work with Java SE :/

Do you see any major problem to just deserialize to untyped Map/List/String/Primitives (outside of an unfriendly untyped api)?

Reply:
There is nothing really that makes this Java EE only. JSON-P spec may be part of Java EE, but both the API and the implementations (in this case Genson itself, which is the only thing I use) can be used in Java SE as well (just like JAX-RS and many other "Java EE" specs).

We can't really deserialize into a plain untyped Map. I use the fact that we are deserializing into `JsonValue` to turn off handling of `@class` attributes, in order to read them like any other JSON property.

I believe that makes sense: once you cross JSON-P barrier, everything should be JSON-P compliant, and we shouldn't construct non-JSON-P classes anywhere in the graph, regardless of `@class` attribute presence.

Reply:
The JSR353 part is currently provided as a bundle, which means that Genson doesn't have a hard requirement on it in order to work. With this change it means that Genson will need to make this dependency mandatory. I've been trying to make sure Genson is a single jar with no external dependency on other libs (for the core Genson to work). I'm not sure I want this to change...

What is your opinion on that?

Why can't we deserialize into a plain untyped Map? Writing a recursive converter that reads a stream to untyped DOM structures (map/array/primitives) should work. You can then use it directly (without going through genson.deserialize(...)) which would ensure that you don't risk of picking class metadata etc.

Reply:
Honestly, I don't think not using JSR353 directly would buy us anything -- we'd basically have to replicate the same functionality, for no good reason.

I hear you on the dependency front, but I also think the line is a bit blurry when it comes to dependency on *standards*, such as JSR353 (JSON-P) and JSR367 (JSON-B, which unfortunately has too many issues in v1 to be usable in real world). I do believe it makes sense to support standards as much as possible, and personally don't see an issue with a hard dependency on the standard API jars.

Reply:
Personally, I feel you are overthinking the whole Java SE vs Java EE thing ;-)

Many Java EE specs and their implementations are perfectly usable from Java SE, and there is no reason we shouldn't treat them as such. The dependencies we are introducing are tiny (only the standard API JARs, really), and could easily become part of Java SE at some point if enough people care about them (chances are they won't, but the reasons for that are more complex than I care to go into and mostly political...)

The real question we need to ask is not which umbrella the spec falls under, SE or EE, but whether it's usable without the Java EE app server or not. The specs like EJB are not, and I would avoid them like a plague (although, I'm not sure what we would even do with them), or definitely keep as separate modules (and separate JARs). However, the specs like JAXB, JSON-P and JSON-B are usable in any Java environment, and are something that, in my opinion, we can use and depend on internally.

Where it gets blurry is with things like JAX-RS. In that case, I would treat it as a separate, optional module, not because it is not usable from Java SE (it certainly is, I use it that way all the time), but because we are implementing its SPI in order to allow it to use Genson for serialization, not the other way around. In other words, JAX-RS extension is not a true Genson bundle -- it is effectively a JAX-RS "bundle/extension" that allows it to integrate with Genson.

Reply:
To make it short I think Genson core shouldn't depend on anything outside of what is already part of the jdk. Be it JAXB/JSONP/JSONB and cie. they only make sense in the Java (EE) world, but for ex. someone that is using the Scala extension has no interest in pulling all that stuff. Same goes for someone that uses it on Android.
I'd like Genson core to remain free of dependencies in order to work properly.